src/parquet/bloom_filter.h

```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  /// responsibility to free the bitset passed in. It is used when reconstructing
+  /// a Bloom filter from a parquet file. It uses murmur3_x64_128 as its default hash
+  /// function and block-based algorithm as default algorithm.
+  ///
+  /// @param bitset The given bitset to construct Bloom filter.
+  /// @param num_bytes  The number of bytes of given bitset.
+  BlockSplitBloomFilter(const uint8_t *bitset, uint32_t num_bytes);
+
+  // Bytes in a tiny Bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // The number of bits to be set in each tiny Bloom filter
+  static constexpr int BITS_SET_PER_BLOCK = 8;
+
+  // A mask structure used to set bits in each tiny Bloom filter.
+  struct BlockMask {
+    uint32_t item[BITS_SET_PER_BLOCK];
+  };
+
+  // The block-based algorithm needs eight odd SALT values to calculate eight indexes
+  // of bit to set, one bit in each 32-bit word.
+  static constexpr uint32_t SALT[BITS_SET_PER_BLOCK] = {
+      0x47b6137bU, 0x44974d91U, 0x8824ad5bU, 0xa2b7289dU,
+      0x705495c7U, 0x2df1424bU, 0x9efc4947U, 0x5c6bfb31U};
+
+  bool FindHash(uint64_t hash) const override;
+  void InsertHash(uint64_t hash) override;
+  void WriteTo(OutputStream *sink) const override;
+  uint32_t GetBitsetSize() const override { return num_bytes_; }
+  uint64_t Hash(int64_t value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(float value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(double value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(const Int96 *value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(const ByteArray *value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(int32_t value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(const FLBA *value, uint32_t len) const override {
     return hasher_->Hash(value, len);
   }
-
-  /// Write this Bloom filter to an output stream. A Bloom filter structure should
-  /// include bitset length, hash strategy, algorithm, and bitset.
+  /// Deserialize the Bloom filter from an input stream. It is used when reconstructing
+  /// a Bloom filter from a parquet filter.
   ///
-  /// @param sink the output stream to write
-  void WriteTo(OutputStream* sink) const;
+  /// @param input_stream The input stream from which to construct the Bloom filter
+  /// @return it returns a Bloom filter pointer point to a new allocated Bloom filter.
+  static BloomFilter *Deserialize(InputStream *input_stream);
 
  private:
+  /// Set bits in mask array according to input key.
+  /// @param key the value to calculate mask values.
+  /// @param mask the mask array is used to set inside a block
+  void SetMask(uint32_t key, BlockMask &mask) const;
+
+  /// Test bits in the bitset according to the hash value.
+  ///
+  /// @param hash the hash used to calculate bit index to test
+  /// @return true
+  bool TestBits(uint64_t hash) const;
```

Comment:
Why have TestBits and SetBits? Why not just put the bodies of those functions in FindHash and InsertHash?